File: php/ProviderWalker.php
<?php

declare(strict_types=1);

/*
 * This file is part of the Valkyrja Framework package.
 *
 * (c) Melech Mizrachi <melechmizrachi@gmail.com>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Valkyrja\Volundr\Analysis;

use Composer\Autoload\ClassLoader;
use PhpParser\Node;
use PhpParser\Node\Expr\Array_;
use PhpParser\Node\Expr\ClassConstFetch;
use PhpParser\Node\Name;
use PhpParser\Node\Stmt\Class_;
use PhpParser\Node\Stmt\ClassMethod;
use PhpParser\Node\Stmt\Interface_;
use PhpParser\NodeFinder;
use PhpParser\NodeTraverser;
use PhpParser\NodeVisitor\NameResolver;
use PhpParser\ParserFactory;
use RuntimeException;

/**
 * Walks a ComponentProviderContract implementation and extracts the FQCNs it
 * returns from each get*Providers method, resolving them to file paths via
 * Composer's autoloader.
 */
final class ProviderWalker
{
    private const METHODS = [
        'componentProviders' => 'getComponentProviders',
        'containerProviders' => 'getContainerProviders',
        'eventProviders'     => 'getEventProviders',
        'cliProviders'       => 'getCliProviders',
        'httpProviders'      => 'getHttpProviders',
    ];

    public function __construct(
        private readonly ClassLoader $loader,
    ) {
    }

    /**
     * @return array{
     *   class: string|null,
     *   componentProviders: list<array{fqcn: string, file: ?string}>,
     *   containerProviders: list<array{fqcn: string, file: ?string}>,
     *   eventProviders:     list<array{fqcn: string, file: ?string}>,
     *   cliProviders:       list<array{fqcn: string, file: ?string}>,
     *   httpProviders:      list<array{fqcn: string, file: ?string}>,
     * }
     */
    public function analyze(string $filePath): array
    {
        $code = file_get_contents($filePath);

        if ($code === false) {
            throw new RuntimeException("Cannot read $filePath");
        }

        $parser = (new ParserFactory())->createForNewestSupportedVersion();
        $ast    = $parser->parse($code) ?? [];

        // Run the NameResolver so `use` statements are applied and every Name
        // node has a resolvedName attribute holding its FQCN.
        $traverser = new NodeTraverser();
        $traverser->addVisitor(new NameResolver());
        $ast = $traverser->traverse($ast);

        $finder = new NodeFinder();

        // Find the class or interface declaration in the file. For a provider
        // we expect exactly one top-level class; interfaces are included so
        // this tool also works against the contract itself for testing.
        /** @var Class_|Interface_|null $decl */
        $decl = $finder->findFirst($ast, static fn (Node $n) => $n instanceof Class_ || $n instanceof Interface_);

        $result = ['class' => $decl?->namespacedName?->toString()];

        foreach (self::METHODS as $key => $_) {
            $result[$key] = [];
        }

        if ($decl === null) {
            return $result;
        }

        foreach (self::METHODS as $key => $methodName) {
            $method = $this->findMethod($decl, $methodName);

            if ($method === null) {
                continue;
            }

            foreach ($this->extractClassStrings($method) as $fqcn) {
                $result[$key][] = [
                    'fqcn' => $fqcn,
                    'file' => $this->resolveToFile($fqcn),
                ];
            }
        }

        return $result;
    }

    private function findMethod(Class_|Interface_ $decl, string $name): ClassMethod|null
    {
        foreach ($decl->getMethods() as $method) {
            if (strcasecmp($method->name->toString(), $name) === 0) {
                return $method;
            }
        }

        return null;
    }

    /**
     * Extract every `Foo::class` reference from the return statements of a
     * method. We only care about literal return arrays — anything dynamic is
     * intentionally not supported, which keeps the invariant clean.
     *
     * @return list<string>
     */
    private function extractClassStrings(ClassMethod $method): array
    {
        $fqcns  = [];
        $finder = new NodeFinder();

        /** @var list<Node\Stmt\Return_> $returns */
        $returns = $finder->findInstanceOf($method->stmts ?? [], Node\Stmt\Return_::class);

        foreach ($returns as $return) {
            if (!$return->expr instanceof Array_) {
                continue;
            }

            foreach ($return->expr->items as $item) {
                if ($item === null || !$item->value instanceof ClassConstFetch) {
                    continue;
                }

                // We only care about `SomeClass::class` (not $var::class).
                if (!$item->value->class instanceof Name) {
                    continue;
                }

                if (!$item->value->name instanceof Node\Identifier) {
                    continue;
                }

                if ($item->value->name->toLowerString() !== 'class') {
                    continue;
                }

                // NameResolver has already attached the fully-qualified name.
                $resolved = $item->value->class->getAttribute('resolvedName');

                if ($resolved instanceof Name) {
                    $fqcns[] = $resolved->toString();
                } else {
                    $fqcns[] = $item->value->class->toString();
                }
            }
        }

        return $fqcns;
    }

    /**
     * Delegate to Composer — it already knows how to resolve PSR-4, PSR-0,
     * and classmap entries, so we don't have to reinvent any of it.
     */
    private function resolveToFile(string $fqcn): string|null
    {
        $path = $this->loader->findFile($fqcn);

        return $path === false ? null : realpath($path) ?: $path;
    }
}

File: go/provider_walker.go
// Package analysis walks a Go implementation of ComponentProviderContract and
// extracts the types returned from each Get*Providers method, resolving each
// to the source file that declares it.
//
// The Go port is assumed to use an interface like:
//
//	type ComponentProvider interface {
//	    GetComponentProviders(app Application) []ComponentProvider
//	    GetContainerProviders(app Application) []ServiceProvider
//	    // ...
//	}
//
// and each implementation returns a literal slice of typed nils or zero values
// (a common Go pattern for type lists), e.g.
//
//	return []ComponentProvider{(*foo.Provider)(nil), (*bar.Provider)(nil)}
//
// We extract those type references and resolve each to a file via the
// packages loader.
package analysis

import (
	"fmt"
	"go/ast"
	"go/token"
	"go/types"
	"strings"

	"golang.org/x/tools/go/packages"
)

// Methods we care about, keyed by the output bucket name.
var methods = map[string]string{
	"componentProviders": "GetComponentProviders",
	"containerProviders": "GetContainerProviders",
	"eventProviders":     "GetEventProviders",
	"cliProviders":       "GetCliProviders",
	"httpProviders":      "GetHttpProviders",
}

// ProviderRef is a single resolved (or unresolved) provider reference.
type ProviderRef struct {
	FQN  string // e.g. "github.com/you/mod/foo.Provider"
	File string // absolute path, or "" if unresolved
}

// Result is the analysis output for one provider file.
type Result struct {
	TypeName  string                   // FQN of the provider type in this file
	Providers map[string][]ProviderRef // keyed by methods[*]
}

// Walker caches a loaded package set so multiple analyze() calls share work.
type Walker struct {
	pkgs []*packages.Package
	fset *token.FileSet
}

// NewWalker loads the given patterns (e.g. "./...") with full type info. The
// loader handles modules, vendor, and GOPATH — we don't reimplement any of it.
func NewWalker(patterns ...string) (*Walker, error) {
	cfg := &packages.Config{
		Mode: packages.NeedName | packages.NeedFiles | packages.NeedSyntax |
			packages.NeedTypes | packages.NeedTypesInfo | packages.NeedImports |
			packages.NeedDeps,
	}
	pkgs, err := packages.Load(cfg, patterns...)
	if err != nil {
		return nil, fmt.Errorf("load packages: %w", err)
	}
	if packages.PrintErrors(pkgs) > 0 {
		return nil, fmt.Errorf("package load had errors")
	}
	return &Walker{pkgs: pkgs, fset: cfg.Fset}, nil
}

// Analyze finds the provider declaration in the given file and extracts the
// class references returned from each Get*Providers method.
func (w *Walker) Analyze(filePath string) (*Result, error) {
	pkg, file := w.findFile(filePath)
	if file == nil {
		return nil, fmt.Errorf("file not found in loaded packages: %s", filePath)
	}

	result := &Result{Providers: make(map[string][]ProviderRef, len(methods))}
	for key := range methods {
		result.Providers[key] = nil
	}

	// Locate the top-level type declaration in this file — we'll use it for
	// the TypeName field and to restrict method lookup to this file's type.
	typeName, methodsByName := collectMethods(file)
	if typeName != "" {
		result.TypeName = pkg.PkgPath + "." + typeName
	}

	for key, methodName := range methods {
		fn, ok := methodsByName[methodName]
		if !ok {
			continue
		}
		refs := w.extractTypeRefs(pkg, fn)
		result.Providers[key] = refs
	}

	return result, nil
}

// findFile locates the *ast.File for a given path across loaded packages.
func (w *Walker) findFile(filePath string) (*packages.Package, *ast.File) {
	for _, pkg := range w.pkgs {
		for i, f := range pkg.CompiledGoFiles {
			if f == filePath {
				return pkg, pkg.Syntax[i]
			}
		}
	}
	return nil, nil
}

// collectMethods walks the file and returns the first concrete type name
// declared in it plus a map of methods defined on that type.
func collectMethods(file *ast.File) (string, map[string]*ast.FuncDecl) {
	methodsByName := map[string]*ast.FuncDecl{}
	var typeName string

	// First pass: find the first non-interface type declaration.
	for _, decl := range file.Decls {
		gd, ok := decl.(*ast.GenDecl)
		if !ok || gd.Tok != token.TYPE {
			continue
		}
		for _, spec := range gd.Specs {
			ts := spec.(*ast.TypeSpec)
			if _, isIface := ts.Type.(*ast.InterfaceType); isIface {
				continue
			}
			typeName = ts.Name.Name
			break
		}
		if typeName != "" {
			break
		}
	}

	// Second pass: gather methods on that type.
	for _, decl := range file.Decls {
		fn, ok := decl.(*ast.FuncDecl)
		if !ok || fn.Recv == nil || len(fn.Recv.List) == 0 {
			continue
		}
		recv := receiverTypeName(fn.Recv.List[0].Type)
		if typeName == "" || recv == typeName {
			methodsByName[fn.Name.Name] = fn
		}
	}
	return typeName, methodsByName
}

func receiverTypeName(expr ast.Expr) string {
	switch t := expr.(type) {
	case *ast.Ident:
		return t.Name
	case *ast.StarExpr:
		return receiverTypeName(t.X)
	}
	return ""
}

// extractTypeRefs walks the method body looking for composite literals inside
// return statements and pulls out the type of each element.
func (w *Walker) extractTypeRefs(pkg *packages.Package, fn *ast.FuncDecl) []ProviderRef {
	var refs []ProviderRef
	if fn.Body == nil {
		return refs
	}

	ast.Inspect(fn.Body, func(n ast.Node) bool {
		ret, ok := n.(*ast.ReturnStmt)
		if !ok {
			return true
		}
		for _, expr := range ret.Results {
			cl, ok := expr.(*ast.CompositeLit)
			if !ok {
				continue
			}
			for _, elt := range cl.Elts {
				if ref := w.refFromExpr(pkg, elt); ref != nil {
					refs = append(refs, *ref)
				}
			}
		}
		return true
	})
	return refs
}

// refFromExpr handles the common "typed nil" pattern used to build type lists
// in Go: (*pkg.Type)(nil). Also handles bare identifiers and zero-value
// composite literals like pkg.Type{}.
func (w *Walker) refFromExpr(pkg *packages.Package, expr ast.Expr) *ProviderRef {
	// (*pkg.Type)(nil) parses as a CallExpr whose Fun is a ParenExpr wrapping
	// a StarExpr. Peel off those layers to find the underlying type expr.
	if call, ok := expr.(*ast.CallExpr); ok {
		if paren, ok := call.Fun.(*ast.ParenExpr); ok {
			if star, ok := paren.X.(*ast.StarExpr); ok {
				return w.refFromTypeExpr(pkg, star.X)
			}
		}
	}
	// pkg.Type{} -> composite literal
	if cl, ok := expr.(*ast.CompositeLit); ok && cl.Type != nil {
		return w.refFromTypeExpr(pkg, cl.Type)
	}
	// Bare identifier (same-package type)
	return w.refFromTypeExpr(pkg, expr)
}

func (w *Walker) refFromTypeExpr(pkg *packages.Package, expr ast.Expr) *ProviderRef {
	tv, ok := pkg.TypesInfo.Types[expr]
	if !ok || tv.Type == nil {
		return nil
	}
	named, ok := tv.Type.(*types.Named)
	if !ok {
		// Might be a pointer — unwrap once.
		if ptr, ok := tv.Type.(*types.Pointer); ok {
			named, ok = ptr.Elem().(*types.Named)
			if !ok {
				return nil
			}
		} else {
			return nil
		}
	}
	obj := named.Obj()
	if obj == nil || obj.Pkg() == nil {
		return nil
	}
	fqn := obj.Pkg().Path() + "." + obj.Name()
	file := w.fset.Position(obj.Pos()).Filename
	return &ProviderRef{FQN: fqn, File: file}
}

// Consume `strings` to keep the import; useful if you extend type-name
// formatting (e.g. trimming vendor prefixes).
var _ = strings.TrimPrefix
